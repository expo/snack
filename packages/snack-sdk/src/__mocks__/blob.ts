export default class Blob {
  static instances: Blob[] = [];

  config: any;

  constructor(config: any) {
    this.config = config;
    Blob.instances.push(this);
  }
}

// @ts-ignore
global.Blob = Blob;

beforeEach(() => {
  Blob.instances = [];
});
