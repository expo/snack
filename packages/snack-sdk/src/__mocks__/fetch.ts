import type { Server } from 'http';

import { defaultConfig } from '..';
import createSnackager from './snackager';
import createVersions from './versions';
import createWww from './www';

// console.log(defaultConfig.apiURL);

const actualFetch = jest.requireActual('node-fetch');

jest.mock('node-fetch', () => {
  return jest.fn(async (url: string, options: Request) => {
    let server: Server | undefined;

    if (!process.env.SNACK_ENV || process.env.SNACK_ENV === 'mock') {
      // Snackager
      if (url.startsWith(defaultConfig.snackagerURL)) {
        const path = url.substring(defaultConfig.snackagerURL.length);
        server = createSnackager().listen(0);
        // @ts-ignore: Property port does not exist on type AddressInfo | string
        url = `http://127.0.0.1:${server.address().port}${path}`;

        // BundledNativeModules.json
      } else if (url.startsWith('https://cdn.jsdelivr.net/npm/expo@')) {
        const path = url.substring('https://cdn.jsdelivr.net/npm'.length);
        server = createVersions().listen(0);
        // @ts-ignore: Property port does not exist on type AddressInfo | string
        url = `http://127.0.0.1:${server.address().port}${path}`;

        // www
      } else if (url.startsWith(defaultConfig.apiURL)) {
        const path = url.substring(defaultConfig.apiURL.length);
        server = createWww().listen(0);
        // @ts-ignore: Property port does not exist on type AddressInfo | string
        url = `http://127.0.0.1:${server.address().port}${path}`;
        // console.log('WWW: ', path);
      } else {
        console.log('UNKNOWN: ', url);
      }
    }

    try {
      return await actualFetch(url, options);
    } catch (e) {
      throw e;
    } finally {
      server?.close();
    }
  });
});

const mockFetch = require('node-fetch');

beforeEach(() => mockFetch.mockClear());

export default mockFetch;
