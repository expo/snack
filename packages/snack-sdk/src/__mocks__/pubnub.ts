import {
  PublishParameters,
  PubnubStatus,
  PublishResponse,
  ListenerParameters,
  SubscribeParameters,
  UnsubscribeParameters,
} from 'pubnub';

const TEST_DEVICES: any = {
  ios: {
    platform: 'ios',
    id: '7682-1234',
    name: 'iPhone 8',
  },
  android: {
    platform: 'android',
    id: 'xxxx-xxxx',
    name: 'Nexus 8',
  },
};

export default class PubNub {
  static instances: PubNub[] = [];
  static devices = TEST_DEVICES;

  config: any;
  listener: ListenerParameters = {};
  channel: string = '';
  publishes: PublishParameters[] = [];
  connectIdx = 0;

  constructor(config: any) {
    this.config = config;
    PubNub.instances.push(this);
  }

  addListener = jest.fn((params: ListenerParameters) => {
    this.listener = params;
  });

  removeListener = jest.fn((_params: ListenerParameters) => {
    this.listener = {};
  });

  subscribe = jest.fn((params: SubscribeParameters) => {
    // @ts-expect-error
    this.channel = params.channels[0];
  });

  unsubscribe = jest.fn((_params: UnsubscribeParameters) => {
    this.channel = '';
  });

  stop = jest.fn(() => {
    this.listener = {};
    this.channel = '';
  });

  publish = jest.fn(
    (
      params: PublishParameters,
      _callback: (status: PubnubStatus, response: PublishResponse) => void
    ) => {
      this.publishes.push(params);
      return Promise.resolve();
    }
  );

  connect(uuid?: string): string {
    const device = TEST_DEVICES[uuid ?? 'ios'];
    uuid = device ? JSON.stringify(device) : uuid;
    // @ts-expect-error
    this.listener.presence({
      action: 'join',
      uuid: uuid ?? '',
      channel: this.channel,
    });
    return uuid ?? '';
  }

  disconnect(uuid?: string, timeout?: boolean) {
    const device = TEST_DEVICES[uuid ?? 'ios'];
    uuid = device ? JSON.stringify(device) : uuid;
    // @ts-expect-error
    this.listener.presence({
      action: timeout ? 'timeout' : 'leave',
      uuid: uuid ?? '',
      channel: this.channel,
    });
    return uuid ?? '';
  }

  sendMessage(message: any, uuid?: string) {
    const device = TEST_DEVICES[uuid ?? 'ios'];
    uuid = device ? JSON.stringify(device) : uuid;
    // @ts-expect-error
    this.listener.message({
      channel: this.channel,
      message,
      publisher: uuid ?? '',
    });
  }
}

beforeEach(() => {
  PubNub.instances = [];
});
