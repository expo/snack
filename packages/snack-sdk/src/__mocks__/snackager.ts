import Router from '@koa/router';
import Koa, { Context } from 'koa';

import defaultConfig, { SnackagerConfig } from '../__fixtures__/snackager';

export default function createSnackager(config: SnackagerConfig = defaultConfig) {
  const app = new Koa();
  const router = new Router();
  function _bundle(ctx: Context, bundleKey: string) {
    try {
      const bundle = config.bundles[bundleKey];
      if (!bundle) throw new Error(`Package '${bundleKey}' not found in the registry`);
      const idx = bundleKey.lastIndexOf('@');
      if (idx < 0) throw new Error('Invalid request, version not specified');
      const name = bundleKey.substring(0, idx);
      const version = bundle.version ?? bundleKey.substring(idx + 1);
      ctx.body = {
        name,
        version,
        hash: `${name}@${version}`,
        handle: `snackager-1/${name.replace('/', '~')}@${version}`,
        dependencies: bundle.peerDependencies,
      };
    } catch (e) {
      ctx.status = 500;
      ctx.body = e.message;
    }
  }
  router.get('/bundle/:key1', (ctx) => _bundle(ctx, ctx.params.key1));
  router.get('/bundle/:key1/:key2', (ctx) => _bundle(ctx, `${ctx.params.key1}/${ctx.params.key2}`));
  app.use(router.routes()).use(router.allowedMethods());
  return app;
}
