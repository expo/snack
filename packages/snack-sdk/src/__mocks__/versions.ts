import Router from '@koa/router';
import Koa from 'koa';

export type VersionsConfig = {
  [key: string]: {
    [key: string]: string;
  };
};

export default function createApp(
  config: VersionsConfig = require('../__fixtures__/bundledNativeModules.json')
) {
  const app = new Koa();
  const router = new Router();
  router.get('/expo@:sdkVersion/bundledNativeModules.json', (ctx) => {
    try {
      const sdkVersion: string = ctx.params.sdkVersion;
      const versions = config[sdkVersion];
      if (!versions) throw new Error(`SDK version '${sdkVersion}' not supported`);
      ctx.body = versions;
    } catch (e) {
      ctx.status = 500;
      ctx.body = e.message;
    }
  });
  app.use(router.routes()).use(router.allowedMethods());
  return app;
}
