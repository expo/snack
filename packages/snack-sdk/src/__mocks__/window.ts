type WindowConfig = {
  location: string;
};

type WindowEventListenerCallback = (event: any) => void;

class WindowLocation {
  public readonly href: string;
  public readonly origin: string;

  constructor(url: string) {
    this.href = url;
    const { origin } = new URL(url);
    this.origin = origin;
  }
}

export default class Window {
  public readonly location: WindowLocation;
  private callback?: (data: any) => void;

  constructor(config: WindowConfig) {
    this.location = new WindowLocation(config.location);
  }

  addEventListener = jest.fn((_type: string, callback: WindowEventListenerCallback) => {
    this.callback = callback;
  });

  removeEventListener = jest.fn((_type: string, _callback: WindowEventListenerCallback) => {
    this.callback = undefined;
  });

  postMessage = jest.fn((message: any, origin: string) => {
    this.callback?.({
      data: message,
      origin,
    });
  });
}

let globalWindowMock: any;

beforeEach(() => {
  globalWindowMock = new Window({
    location: 'https://snack.expo.dev',
  });
  Object.keys(globalWindowMock).forEach((key) => {
    // @ts-ignore
    global[key] = globalWindowMock[key];
  });
});

export function postMessage(message: any, origin?: string) {
  globalWindowMock.postMessage(message, origin ?? 'https://snack.expo.dev');
}
