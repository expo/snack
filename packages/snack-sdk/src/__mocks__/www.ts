import Router from '@koa/router';
import Koa from 'koa';
import bodyParser from 'koa-bodyparser';

import type { SnackFiles } from '..';

export default function createWww() {
  const app = new Koa();
  app.use(bodyParser());
  const router = new Router();
  router.post('/--/api/v2/snack/save', (ctx) => {
    try {
      // @ts-ignore
      const code: SnackFiles = ctx.request.body.code;
      if (!code['App.js'] && !code['App.tsx']) {
        throw new Error('Invalid entry point');
      }
      const key = '@snack-sdk/test';
      ctx.body = {
        id: key,
        key,
        hashId: '00000000',
      };
    } catch (e) {
      ctx.status = 500;
      ctx.body = e.message;
    }
  });
  router.post('/--/api/v2/snack/uploadAsset', (ctx) => {
    const hash = 'FDFDFDFDFDFDFD';
    ctx.body = {
      url: `https://snack-code-uploads.s3.us-west-1.amazonaws.com/~asset/${hash}`,
      hash,
    };
  });
  router.get('/--/api/v2/versions/latest', (ctx) => {
    ctx.body = {
      data: {
        sdkVersions: {
          '44.0.0': {
            iosClientUrl: 'https://dpq5q02fu5f55.cloudfront.net/Exponent-2.23.2.tar.gz',
            releaseNoteUrl: 'https://blog.expo.dev/expo-sdk-44-4c4b8306584a',
            relatedPackages: {
              jest: '^26.6.3',
              typescript: '~4.3.5',
              '@babel/core': '^7.12.9',
              '@types/react': '~17.0.21',
              '@types/react-dom': '~17.0.9',
              'react-native-web': '0.17.1',
              'babel-preset-expo': '9.0.2',
              '@types/react-native': '~0.64.12',
              '@expo/webpack-config': '~0.16.2',
              'react-native-unimodules': '~0.15.0',
            },
            androidClientUrl: 'https://d1ahtucjixef4r.cloudfront.net/Exponent-2.23.2.apk',
            iosClientVersion: '2.23.2',
            expoReactNativeTag: 'sdk-44.0.0',
            androidClientVersion: '2.23.2',
            facebookReactVersion: '17.0.1',
            facebookReactNativeVersion: '0.64.3',
            packagesToInstallWhenEjecting: {
              'react-native': 'https://github.com/expo/react-native/archive/sdk-44.0.0.tar.gz',
              'react-native-unimodules': '0.15.0',
            },
          },
          '45.0.0': {
            iosClientUrl: 'https://dpq5q02fu5f55.cloudfront.net/Exponent-2.24.3.tar.gz',
            releaseNoteUrl: 'https://blog.expo.dev/expo-sdk-45-f4e332954a68',
            relatedPackages: {
              jest: '^26.6.3',
              typescript: '~4.3.5',
              '@babel/core': '^7.12.9',
              '@expo/config': '~6.0.19',
              '@types/react': '~17.0.21',
              '@types/react-dom': '~17.0.11',
              'react-native-web': '0.17.7',
              'babel-preset-expo': '~9.1.0',
              '@types/react-native': '~0.67.6',
              '@expo/config-plugins': '^4.1.0',
              '@expo/webpack-config': '~0.16.21',
              '@expo/prebuild-config': '^4.0.0',
              'expo-modules-autolinking': '~0.8.1 || ~0.9.0',
            },
            androidClientUrl: 'https://d1ahtucjixef4r.cloudfront.net/Exponent-2.24.6.apk',
            iosClientVersion: '2.24.3',
            expoReactNativeTag: 'sdk-45.0.0',
            androidClientVersion: '2.24.6',
            facebookReactVersion: '17.0.2',
            facebookReactNativeVersion: '0.68.2',
            packagesToInstallWhenEjecting: {
              'react-native': 'https://github.com/expo/react-native/archive/sdk-45.0.0.tar.gz',
              'react-native-unimodules': '0.15.0',
            },
          },
          '46.0.0': {
            beta: 'true',
            iosClientUrl: 'https://dpq5q02fu5f55.cloudfront.net/Exponent-2.25.1.tar.gz',
            relatedPackages: {
              jest: '^26.6.3',
              typescript: '^4.6.3',
              '@babel/core': '^7.18.6',
              '@expo/config': '^7.0.0',
              '@types/react': '~18.0.0',
              '@types/react-dom': '~18.0.0',
              'react-native-web': '~0.18.7',
              'babel-preset-expo': '~9.2.0',
              '@types/react-native': '~0.69.1',
              '@expo/config-plugins': '^5.0.0',
              '@expo/webpack-config': '^0.17.0',
              '@expo/prebuild-config': '^5.0.1',
              'expo-modules-autolinking': '~0.10.1',
            },
            androidClientUrl: 'https://d1ahtucjixef4r.cloudfront.net/Exponent-2.25.1.apk',
            iosClientVersion: '2.25.1',
            expoReactNativeTag: 'sdk-46.0.0',
            androidClientVersion: '2.25.1',
            facebookReactVersion: '18.0.0',
            facebookReactNativeVersion: '0.69.3',
            packagesToInstallWhenEjecting: {
              'react-native': 'https://github.com/expo/react-native/archive/sdk-46.0.0.tar.gz',
              'react-native-unimodules': '0.15.0',
            },
          },
        },
      },
    };
  });
  app.use(router.routes()).use(router.allowedMethods());
  return app;
}
