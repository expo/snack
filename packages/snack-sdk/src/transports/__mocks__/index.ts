import { SnackTransport, SnackTransportOptions } from '../index';
import TestTransport from './TestTransport';

export function createTransport(options: SnackTransportOptions): SnackTransport {
  const result = new TestTransport(options);
  return result;
}

export function createSnackPubTransport(options: SnackTransportOptions): SnackTransport {
  return new TestTransport(options);
}
