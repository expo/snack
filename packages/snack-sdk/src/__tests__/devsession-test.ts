import fetch from 'node-fetch';

import '../__mocks__/blob';
import Snack from './snack-sdk';

jest.mock('node-fetch');

// @ts-ignore
fetch.mockReturnValue(
  Promise.resolve({
    ok: true,
    status: 200,
    json: async () => ({
      data: {
        url: 'whoop',
      },
    }),
  })
);

beforeEach(() => {
  // @ts-ignore
  fetch.mockClear();
});

describe('devsession', () => {
  it('sends notify when online and user or device-id specified', async () => {
    const snack = new Snack({});
    expect(fetch).not.toBeCalled(); // !online
    snack.setOnline(true);
    expect(fetch).not.toBeCalled(); // !device-id
    snack.setOnline(false);
    expect(fetch).not.toBeCalled(); // !online
    snack.setOnline(true);
    snack.setDeviceId('1234');
    expect(fetch).toBeCalledTimes(1); // online && device-id
    snack.setOnline(false);
    expect(fetch).toBeCalledTimes(2); // close
    snack.setDeviceId();
    expect(fetch).toBeCalledTimes(2); // still closed
    snack.setOnline(true);
    expect(fetch).toBeCalledTimes(2); // !device-id
    snack.setDeviceId('1234');
    expect(fetch).toBeCalledTimes(3); // notify
    snack.setOnline(false);
    expect(fetch).toBeCalledTimes(4); // !online
  });

  it('receives sendBeaconCloseRequest', async () => {
    const snack = new Snack({
      apiURL: 'https://exp.host',
      channel: '10spnBnPxi',
    });
    expect(snack.getState().sendBeaconCloseRequest).toBeUndefined();
    expect(fetch).not.toBeCalled();
    snack.setOnline(true);
    expect(fetch).not.toBeCalled();
    snack.setDeviceId('1234');
    expect(fetch).toBeCalledTimes(1);
    await new Promise((resolve) => setTimeout(resolve, 1));
    expect(snack.getState().sendBeaconCloseRequest).toBeDefined();
    expect(snack.getState().sendBeaconCloseRequest).toMatchSnapshot();
    snack.setOnline(false);
    expect(fetch).toBeCalledTimes(2);
  });

  it('sends notify when calling setFocus', async () => {
    const snack = new Snack({
      apiURL: 'https://exp.host',
      online: true,
      deviceId: '1234',
    });
    expect(fetch).toBeCalledTimes(1);
    snack.setFocus();
    expect(fetch).toBeCalledTimes(2);
    snack.setOnline(false);
    expect(fetch).toBeCalledTimes(3);
  });

  it('sends notify with user session secret', async () => {
    const snack = new Snack({
      apiURL: 'https://exp.host',
      online: true,
      deviceId: '1234',
      user: { sessionSecret: '{"some":"json"}' },
    });
    expect(fetch).toBeCalledWith(
      expect.any(String),
      expect.objectContaining({
        headers: expect.objectContaining({
          'Expo-Session': '{"some":"json"}',
        }),
      })
    );
    snack.setOnline(false);
  });

  it('sends notify with user access token', async () => {
    const snack = new Snack({
      apiURL: 'https://exp.host',
      online: true,
      deviceId: '1234',
      user: { accessToken: 'sometoken' },
    });
    expect(fetch).toBeCalledWith(
      expect.any(String),
      expect.objectContaining({
        headers: expect.objectContaining({
          Authorization: 'Bearer sometoken',
        }),
      })
    );
    snack.setOnline(false);
  });
});
