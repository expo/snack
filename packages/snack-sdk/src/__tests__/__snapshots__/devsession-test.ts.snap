// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`devsession receives sendBeaconCloseRequest 1`] = `
Object {
  "data": Blob {
    "config": Array [
      "{\\"session\\":{\\"url\\":\\"exp://exp.host/@snack/sdk.42.0.0-10spnBnPxi\\"}}",
    ],
  },
  "url": "https://exp.host/--/api/v2/development-sessions/notify-close?deviceId=1234",
}
`;
